libass/ass_render.c

```diff
@@ -1495,25 +1497,43 @@ static void measure_text(ASS_Renderer *render_priv)
     int cur_line = 0;
     double scale = 0.5 / 64;
     int max_asc = 0, max_desc = 0;
+    double max_border_y = 0, max_border_x = 0;
     for (int i = 0; i < text_info->length; i++) {
         if (text_info->glyphs[i].linebreak) {
             max_asc = max_desc = 0;
+            max_border_y = 0;
             scale = 0.5 / 64;
             cur_line++;
         }
         GlyphInfo *cur = text_info->glyphs + i;
         max_asc  = FFMAX(max_asc,  cur->asc);
         max_desc = FFMAX(max_desc, cur->desc);
+        max_border_y = FFMAX(max_border_y, cur->border_y);
+        max_border_x = FFMAX(max_border_x, cur->border_x);
         if (cur->symbol != '\n' && cur->symbol != 0)
             scale = 1.0 / 64;
         if (i == text_info->length - 1 || text_info->glyphs[i + 1].linebreak) {
             text_info->lines[cur_line].asc  = scale * max_asc;
             text_info->lines[cur_line].desc = scale * max_desc;
             text_info->height += scale * max_asc + scale * max_desc;
+            // For *VSFilter compatibility do biased rounding on max_border*
+            // https://github.com/Cyberbeing/xy-VSFilter/blob/xy_sub_filter_rc4@%7B2020-05-17%7D/src/subtitles/RTS.cpp#L1465
+            text_info->border_bottom =
+                (int) (render_priv->border_scale * max_border_y + 0.5);
```

Comment:
C99 have `lrint()` for such rounding.

Reply:
I think this is rather `lround` (when the argument is nonnegative, which I assume it is).

Also note that we have many more instances of this that could use the C99 functions but don’t. But sure, we can aim to use them in new code and maybe convert old code eventually.

Reply:
`lrint` usually converts to more optimized assembly than `lround` and in practice there is no difference between results.

Reply:
What do you mean “there is no difference in results”? They do different things. As the comment explicitly says here, we want `lround`.

Reply:
They do the same in default rounding mode except some edge cases.

Reply:
The “edge cases” you speak of are precisely the entire point. They round 1.5 an 2.5 differently.

Reply:
`grep`ping for `'+ 0.5'`, most instances of rounding like this are already touched by this commit and could be changed to `lround`, assuming it is safe to assume `>=0` in the respective case.
In case of the `max_border_…`s `>=0` is guaranteed to be true.  
In case of `event_image->left` this is **not** guaranteed  
*(but I don't know if `+0.5` or `lround` would match VSFilter; if it's not `lround/lrint` this would create a cosmetic local inconsistency with `width` and `height` using `lround/lrint` and `left` using `+0.5`, but feel free to disregard this)*

`lround` is currently not being used in libass. 
`lrint` is used in `ass_blur.c` and `ass_outline.c`.

The behaviour of lrint depends on `fe{s,g}etround()` though. libass doesn't seem to set it to any particular value and just uses the machines default. I haven't seen a mention of a standard default mode, but it's possible I missed that.
Also when setting the mode; quoting C99-draft:
> [each of the rounding mode macros] is defined if and only if the implementation supports getting and setting the represented
rounding direction by means of the fegetround and fesetround functions.
[note: and `FE_TONEAREST` seems to do **un**biased rounding]

*(POSIX probably guarantees some to exist, haven't checked though)*

What assumptions does libass make about the target machine besides supporting C99 btw ?

Reply:
Nobody changes rounding mode in 2020. It's always safe to assume nearest rounding mode.

Reply:
I agree with both of you here. 😅

On the one hand, it’s true that technically `lrint` actually isn’t … portable? fully defined? reliable? whatever you want to call it, if we don’t set or check the rounding mode. But on the other hand, here, indeed, I think we can say that in practice it’s always round-to-nearest-and-half-to-even.

At most, it might not be on some relatively exotic hardware platforms, but I can’t find anything with a quick Google search. One source also says, “The IEEE standard requires that the default rounding mode be round to nearest.” and I think we kinda already assume IEEE floating point (in `ass_strtod` in particular), although, of course, whatever software is calling libass could change the rounding mode before invoking our functions.

I think it’s safe to assume `lrint` and primitive floating-point arithmetic operations do half-to-even while `lround` does half-away-from-zero. But I do think the rounding mode is a big deal, and substituting `lrint` for `lround` is wrong.

As for more efficient machine code (even though saving a couple assembler instructions per character while knowingly sacrificing correctness doesn’t seem to make sense to me anyway), without `-ffast-math`, both `lround` and `lrint` [are compiled](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(j:1,lang:___c,source:'%23include+%3Cstdio.h%3E%0A%23include+%3Cmath.h%3E%0A%0Atypedef+union+%7B%0A%09long+long+i%3B%0A%09double+f%3B%0A%7D+foo%3B%0A%0A__attribute__((noinline))%0Avoid+func(double*+f,+long+long*+i)+%7B%0A++++double+a+%3D+*f%3B%0A++++*i+%3D+4%3B%0A++++double+b+%3D+*f%3B%0A++++printf(%22%25f+%25f%5Cn%22,+a,+b)%3B%0A%7D%0A%0Avoid+func_caller()+%7B%0A++++foo+f%3B%0A++++func(%26f.f,+%26f.i)%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang600,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',trim:'1'),lang:___c,libs:!(),options:'-O3',source:1),l:'5',n:'0',o:'x86-64+clang+6.0.0+(Editor+%231,+Compiler+%231)+C',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) into actual library function calls. Even with `-ffast-math`, [only GCC](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(j:1,lang:___c,source:'%23include+%3Cstdio.h%3E%0A%23include+%3Cmath.h%3E%0A%0Atypedef+union+%7B%0A%09long+long+i%3B%0A%09double+f%3B%0A%7D+foo%3B%0A%0A__attribute__((noinline))%0Avoid+func(double*+f,+long+long*+i)+%7B%0A++++double+a+%3D+*f%3B%0A++++*i+%3D+4%3B%0A++++double+b+%3D+*f%3B%0A++++printf(%22%25f+%25f%5Cn%22,+a,+b)%3B%0A%7D%0A%0Avoid+func_caller()+%7B%0A++++foo+f%3B%0A++++func(%26f.f,+%26f.i)%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang600,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',trim:'1'),lang:___c,libs:!(),options:'-O3',source:1),l:'5',n:'0',o:'x86-64+clang+6.0.0+(Editor+%231,+Compiler+%231)+C',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) (not Clang, not Intel, [not MSVC](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(j:1,lang:___c,source:'%23include+%3Cstdio.h%3E%0A%23include+%3Cmath.h%3E%0A%0Atypedef+union+%7B%0A%09long+long+i%3B%0A%09double+f%3B%0A%7D+foo%3B%0A%0A__attribute__((noinline))%0Avoid+func(double*+f,+long+long*+i)+%7B%0A++++double+a+%3D+*f%3B%0A++++*i+%3D+4%3B%0A++++double+b+%3D+*f%3B%0A++++printf(%22%25f+%25f%5Cn%22,+a,+b)%3B%0A%7D%0A%0Avoid+func_caller()+%7B%0A++++foo+f%3B%0A++++func(%26f.f,+%26f.i)%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang600,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',trim:'1'),lang:___c,libs:!(),options:'-O3',source:1),l:'5',n:'0',o:'x86-64+clang+6.0.0+(Editor+%231,+Compiler+%231)+C',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)) substitutes builtins and even then `lrint` is only one instruction shorter than `(int) (x + 0.5)`, although `lround` is indeed considerably longer.

So if speed is (somehow) important here, the current code, `(int) (x + 0.5)`, is actually best.

And if we need the `+ 0.5` in a case where numbers can be negative and in the next line we have numbers that can’t be negative, it would be more consistent/easier to read to use `(int) (x + 0.5)` in both cases. I remember explicitly using `+ 0.5` with negative numbers in… parsing, maybe? for VSFilter compatibility.

> What assumptions does libass make about the target machine besides supporting C99 btw ?

We definitely assume `int` holds 32 bits, and possibly that it’s two’s complement. We probably assume some amount of IEEE float compatibility (especially in `ass_strtod`). We assume obscure things about integer types like: the sign bit of a signed type isn’t a padding bit in the corresponding unsigned type. There was some talk of this in #261 when I tried to make code handle various odd cases allowed by the standard and was shot down.

Reply:
Sorry, my Godbolt links were broken. Check the updated comment for the right ones.

Reply:
And speaking of edge cases, `(int) (x + 0.5)` differs from `lround(x)` (_and_ `lrint(x)`)
* if `x` is the largest representable number smaller than 0.5, namely 0.49999999999999994 (the naïve code rounds to 1, while `lround` to 0),
* or if `x` is any odd 53-bit integer (the naïve code adds 1, while `lround` preserves the same value).

Admittedly, the 53-bit integer case shouldn’t be very relevant to us given that we cast the whole thing to a 32-bit `int`…

All in all, I’d prefer us to stick to exactly the same code that VSFilter uses where it matters, and to use `lrint` where we want actually good rounding.